addons/binding/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/handler/TeslaHandler.java

```diff
@@ -0,0 +1,948 @@
+	protected VehicleState vehicleState;
+	protected ChargeState chargeState;
+	protected ClimateState climateState;
+	protected String accessToken;
+
+	// REST Client API variables
+	protected Client teslaClient = ClientBuilder.newClient();
+	protected Client eventClient;
+	protected WebTarget teslaTarget = teslaClient.target(TESLA_OWNERS_URI);
+	protected WebTarget tokenTarget = teslaTarget.path(TESLA_ACCESS_TOKEN_URI);
+	protected WebTarget vehiclesTarget = teslaTarget.path(API_VERSION).path(
+			VEHICLES);
+	protected WebTarget vehicleTarget = vehiclesTarget.path(VEHICLE_ID_PATH);
+	protected WebTarget dataRequestTarget = vehicleTarget
+			.path(DATA_REQUEST_PATH);
+	protected WebTarget commandTarget = vehicleTarget.path(COMMAND_PATH);
+	protected WebTarget eventTarget;
+
+	// Threading and Job related variables
+	protected final ScheduledThreadPoolExecutor scheduler = (ScheduledThreadPoolExecutor) Executors
+			.newScheduledThreadPool(10);
+	protected ScheduledFuture<?> eventJob;
+	protected ScheduledFuture<?> fastStateJob;
+	protected ScheduledFuture<?> slowStateJob;
+	protected QueueChannelThrottler stateThrottler;
+	protected Response eventResponse;
+	protected BufferedReader eventBufferedReader;
+	protected InputStreamReader eventInputStreamReader;
+	protected long lastEventSystemTime = 0;
+	protected String lastEventTimeStamp = "";
+
+	public TeslaHandler(Thing thing) {
+		super(thing);
+	}
+
+	@Override
+	public void initialize() {
+
+		if (!authenticate((String) getConfig().get(USERNAME),
+				(String) getConfig().get(PASSWORD))) {
+			logger.warn("Unable to connect with the supplied credentials");
+			updateStatus(ThingStatus.OFFLINE);
+		} else {
+
+			// get a list of vehicles
+			Response response = vehiclesTarget
+					.request(MediaType.APPLICATION_JSON_TYPE)
+					.header("Authorization", "Bearer " + accessToken).get();
+
+			if (response != null && response.getStatus() == 200
+					&& response.hasEntity()) {
+
+				if ((vehicle = queryVehicle()) != null) {
+
+					logger.debug(
+							"Found the vehicle with VIN '{}' in the list of vehicles you own",
+							(String) getConfig().get(VIN));
+					updateStatus(ThingStatus.ONLINE);
+
+					if (eventJob == null || eventJob.isCancelled()) {
+						eventJob = scheduler.scheduleAtFixedRate(eventRunnable,
+								0, EVENT_REFRESH_INTERVAL,
+								TimeUnit.MILLISECONDS);
+					}
+
+					Map<Object, Rate> channels = new HashMap<Object, Rate>();
+					channels.put(TESLA_DATA_THROTTLE, new Rate(10, 10,
+							TimeUnit.SECONDS));
+					channels.put(TESLA_COMMAND_THROTTLE, new Rate(20, 1,
+							TimeUnit.MINUTES));
+
+					Rate firstRate = new Rate(20, 1, TimeUnit.MINUTES);
+					Rate secondRate = new Rate(200, 10, TimeUnit.MINUTES);
+					stateThrottler = new QueueChannelThrottler(firstRate,
+							scheduler, channels);
+					stateThrottler.addRate(secondRate);
+
+					if (fastStateJob == null || fastStateJob.isCancelled()) {
+						fastStateJob = scheduler.scheduleAtFixedRate(
```

Comment:
No. But it would not make a lot of difference since the fastStateRunnable does not lot a do apart from scheduling other Runnables with the stateThrottler.